### DISCUSSION TITLE
Navegação lenta para rota

### DISCUSSION THREAD

**Discussion by GCoeCta** [balta-io/forum#469] [Tira Dúvidas] [answered]
Caros,

Estou criando minha primeira aplicação Blazor a partir dos cursos do André.

Nesta aplicação inicial tenho uma página inicial na qual o operador realiza pesquisas em um banco de dados com cadastros de fazendas. A pesquisa é feita a partir de uma string com o início do nome das fazendas a pesquisa ( exemplo "Fazenda Santa" , "Fazenda Sao" que trarão resultados como Fazenda Santa Maria, Fazenda Santa Clara etc ) e o estado onde a fazenda está. 

São retornados 20 registros por pesquisa.

Quando o operador aciona o comando "Pesquisar agora" na página inicial, ele navega para uma outra página na qual é feita a pesquisa e os resultados são apresentados. A parte inicial do nome a ser pesquisado e a sigla do estado são passados a esta segunda página como parâmetros.

Nesta segunda tela estou usando @attribute [StreamRendering(true)] de forma a que antes dos resultados serem mostrados, seja apresentado um spinner indicando que o processo está em andamento.

Acontece que após o comando de pesquisa ser acionado na tela inicial, ao invés de imediatamente ser apresentada a segunda tela, o programa permanece na primeira enquanto o processamento ocorre na segunda. Assim, acionamos o comando de pesquisa, a rota é ajustada no navegador mas a tela não muda. 

Para contornar esta situação coloquei na primeira tela um spinner acionado pelo comando de pesquisa.

Alguma idéia do que pode estar ocorrendo ?

Abaixo o link para esta aplicação.

[https://blazorapp01.azurewebsites.net/](url)

Obrigado

**Comment by andrebaltieri** [comment, marked as answer]
Boa tarde, @GCoeCta, como vai?

Preciso ver o código que está fazendo (Cuidado não expor nada sensível) para poder entender melhor o que está ocorrendo!
Parece algo mais comportamental do que um bug!

Posta os códigos que conseguimos ajudar melhor.

**Comment by GCoeCta** [comment]
Boa tarde André !

Segue aqui a página índex que é o ponto de entrada da aplicação e a página seguinte, onde os resultados são apresentados ( e que deveria ser apresentada de imediato após o acionamento do processamento ) : 

`@rendermode InteractiveServer
@inject NavigationManager Navigation
@using System.ComponentModel.DataAnnotations


@page "/"
<div class="container">
    <div class="row vh-100 justify-content-center align-items-center w-25 mx-auto">

        <div class="bg-light rounded-2 p-2">
            <h4 class="mb-4 text-center shadow p-3 rounded-3 text-white text-xl bg-success">Pesquisar fazenda</h4>

            @if (Processando == true)
            {
                <div class="text-center">
                    <div class="spinner-border text-success" role="status">
                    </div>
                </div>
            }


            <EditForm Model=@Model OnValidSubmit=@Processar>
                <DataAnnotationsValidator />
                <ValidationSummary />
                <div class="form-group">
                    <label for="Nome">Nome a pesquisar</label>

                    <InputText class="form-control mb-2" id="Nome" @bind-Value="Model.NomeDaFazenda"> </InputText>

                </div>
                <div class="form-group">

                    <label for="SelectEstados">Estado</label>

                    <InputSelect id="SelectEstados" class="form-control mb-2" @bind-Value="Model.UF">

                        <option value="AC">AC</option>
                        <option value="AL">AL</option>
                        <option value="AP">AP</option>
                        <option value="AM">AM</option>
                        <option value="BA">BA</option>
                        <option value="CE">CE</option>
                        <option value="DF">DF</option>
                        <option value="ES">ES</option>
                        <option value="GO">GO</option>
                        <option value="MA">MA</option>
                        <option value="MT">MT</option>
                        <option value="MS">MS</option>
                        <option value="MG">MG</option>
                        <option value="PA">PA</option>
                        <option value="PB">PB</option>
                        <option value="PR">PR</option>
                        <option value="PE">PE</option>
                        <option value="PI">PI</option>
                        <option value="RJ">RJ</option>
                        <option value="RN">RN</option>
                        <option value="RS">RS</option>
                        <option value="RO">RO</option>
                        <option value="RR">RR</option>
                        <option value="SC">SC</option>
                        <option value="SP">SP</option>
                        <option value="SE">SE</option>
                        <option value="SC">SC</option>
                        <option value="TO">TO</option>

                    </InputSelect>
                </div>

                <div class="text-center">
                    <button type="submit" class="btn btn-primary mt-4" style="width:200px">
                        <i class="fa fa-home p-2"></i>Pesquisar agora
                    </button>
                </div>

            </EditForm>

        </div>

    </div>

</div>


@code {

    private string UF = "";
    private bool Processando = false;

    private void Processar()
    {
        Processando = true;
        Navigation.NavigateTo("/fazendas/" + Model.NomeDaFazenda + "/" + Model.UF);
    }


    private class MyModel
    {
        [Required(ErrorMessage = "Informe o início do nome pelo menos")]
        public string NomeDaFazenda { get; set; } = string.Empty;
        [Required(ErrorMessage = "Informe o estado")]
        public string UF { get; set; } = string.Empty;
    }

    MyModel Model = new MyModel();


}`


Página dos resultados :  

`@page "/fazendas/{nome}/{UF}"
@attribute [StreamRendering(true)]
@inject appContext dbFazendas ;
@inject NavigationManager Navegator;



<PageTitle>Cadastro de Fazendas do Brasil </PageTitle>


@if ((ListaDeFazendas is not null) && ProcessamentoEncerrado)
{

    if (ListaDeFazendas.Any())
    {

        <div class="col-7 m-auto">
            <BotaoHome />

            <h2 class="mb-4 text-center bg-success rounded-3 shadow-lg p-2 text-white">Cadastro Nacional de Fazendas</h2>

            <table class="table table-striped table-hover">

                <thead>

                    <tr>
                        <th>Fazenda</th>
                        <th>Proprietario</th>
                        <th>Área</th>
                        <th>UF</th>
                        <th>Município</th>
                    </tr>

                </thead>

                <tbody>
                    @foreach (var Fazenda in ListaDeFazendas)
                    {
                        <tr>
                            <td><a href="/fazenda/@Fazenda.Id">
                                @Fazenda.NomeDoImovel
                                </a>
                            </td>
                            <td>
                                @Fazenda.Titular
                            </td>
                            <td>
                                @Fazenda.AreaTotal
                            </td>
                            <td>
                                @Fazenda.Uf
                            </td>
                            <td>
                                @Fazenda.Municipio
                            </td>

                        </tr>
                    }
                </tbody>

            </table>

        </div>


    }
    else
    {
        <div class="text-center">
            <h1>Não há dados !</h1>
        </div>

    }

}
else
{

    <div class="text-center">
        <div class="spinner-border text-success" role="status">
        </div>
    </div>

}


@code {

    [Parameter]
    public string? Nome { get; set; }

    [Parameter]
    public string? UF { get; set; }

    public int Registros = 0;
    public bool ProcessamentoEncerrado = false;
    public IEnumerable<Sncr> ListaDeFazendas = Enumerable.Empty<Sncr>();
    public string Error = "aaaa";



    protected override async Task OnInitializedAsync()
    {
        try
        {

            dbFazendas.Database.SetCommandTimeout(TimeSpan.FromMinutes(20));
            ListaDeFazendas = await dbFazendas.Sncrs.Where(x => x.NomeDoImovel.StartsWith(@Nome) && x.Uf == UF).Take(100).ToListAsync();

            ProcessamentoEncerrado = true;
            if (ListaDeFazendas.Any())
            {
                // Some code...
            }

        }
        catch (Exception x)
        {

            Error = x.Message;

        }
    }

}

Obrigado


`

**Reply by andrebaltieri** [reply]
@GCoeCta e qual comportamento acontece em cada etapa? Você consegue explicar os trechos de código? O comportamento atual e o que gostaria que acontecesse?

**Comment by GCoeCta** [comment]
Oi André,

O que ocorre é que na tela inicial, se no nome da fazenda entro com "Fazenda Santa" em UF -> MT, de forma a encontrar em Mato Grosso todas as fazendas cujo nome comece com "Fazenda Santa..." ao iniciar a pesquisa, a rota fica definida e apresentada no navegador mas ao invés de imediatamente a aplicação passar a segunda página apresentando a indicação de processamento ( o spinner ), a aplicação fica na primeira página ( apresentando um spinner que acrescentei para contornar o problema ) ao invés de já apresentar a página para onde aponta a rota. 

Este comportamente vc pode conferir na aplicação -> https://blazorapp01.azurewebsites.net/

**Comment by andrebaltieri** [comment]
@GCoeCta como vai? Fiz um teste aqui e me parece que o comportamento da sua página inicial está correta.. ela seta o Processando e navega para outra página.

Porém, a outra página na qual ela navega, está levando mais de um minuto para carregar e depois exibe o texto:
https://blazorapp01.azurewebsites.net/fazendas/teste/MG

![Screenshot 2024-01-20 at 09 12 13](https://github.com/balta-io/forum/assets/965305/0d4d2797-377a-4628-8928-55618902dde4)
G

Em resumo, seu código está certo, só que sua consulta realmente demora e como você faz ela no OnInicialized, ela espera os dados para completar.

Esta aula vai te ajudar a resolver este problema:
https://balta.io/cursos/fundamentos-blazor-dotnet-8/3002M01A11
